aten/src/ATen/native/mkldnn/xpu/Blas.cpp

```diff
@@ -512,4 +513,52 @@ TORCH_IMPL_FUNC(addmv_out_xpu)
   xpu::addmv_out(self, mat, vec, beta, alpha, const_cast<Tensor&>(result));
 }
 
+
+Tensor _weight_int4pack_mm_xpu(
+    const Tensor& A,
+    const Tensor& B,
+    int64_t qGroupSize,
+    const Tensor& qScale,
+    const Tensor& qZeros) {
+
+  constexpr int64_t kNTileSize = 8;
+
+  auto M = A.size(0); // M
+  auto N = B.size(0) * kNTileSize;  // N1=LCM(N, K)
+  auto K = A.size(1); // K1=LCM(K, 1024)
+  TORCH_CHECK(A.dtype() == kBFloat16 || A.dtype() == kHalf || A.dtype() == kFloat,
+      __func__, " : expect A to be either 32-bit or 16-bit float tensor.");
+  TORCH_CHECK(A.is_contiguous(),
+      __func__, " : expect A to be contiguous.");
+  TORCH_CHECK(A.dim() == 2,
+      __func__, " : expect A to be 2D tensor.");
+
+  TORCH_CHECK(B.dtype() == kInt,
+      __func__, " : expect B to be int32 tensor.");
+  TORCH_CHECK(B.dim() == 4,
+      __func__, " : expect B to 4d tensor.");
+
+  TORCH_CHECK(qGroupSize == 32 || qGroupSize == 64 || qGroupSize == 128
+      || qGroupSize == 256,
+      __func__, ": expect qGroupSize to be 32, 64, 128 or 256, got ", qGroupSize);
+
+  TORCH_CHECK(qScale.dim() == 2 && qScale.size(1) == N, 
+      __func__, ": expect qScale to be 2d tensor with sizes [:, ", N, "]");
+    TORCH_CHECK(qZeros.dim() == 2 && qZeros.size(1) == N,
+      __func__, ": expect qZeros to be 2d tensor with sizes [:, ", N, "]");
+  
+  auto C = at::empty({M, N}, A.options());
+  
+  //qScaleAndZeros [K/qGroupSize, N, 2]
+  //qscale:[K/qGroupSize, N]
+  //qzp:[K/qGroupSize, N]
+  // auto qscale = qScaleAndZeros.select(2, 0).contiguous();
+  // auto qzp = qScaleAndZeros.select(2, 1).contiguous().to(kChar);
+  std::cout<<"At "<<__FILE__<<" Line:"<<__LINE__<<std::endl;
```

Comment:
The zp and scale are different tensors now. Pls remove these comments.

Reply:
These comments are removed now, thanks for reminding.